src/data/finset/powerset.lean

```diff
@@ -67,8 +67,63 @@ begin
     simp [finset.erase_eq_of_not_mem h, this] }
 end
 
+/-- For predicate `p` decidable on subsets, it is decidable whether `p` holds for any subset. -/
+def decidable_exists_of_decidable_subsets {s : finset α} {p : ∀ t ⊆ s, Prop}
+  (hu : ∀ t (h : t ⊆ s), decidable (p t h)) : decidable (∃ t (h : t ⊆ s), p t h) :=
+decidable_of_iff (∃ t (hs : t ∈ s.powerset), p t (mem_powerset.1 hs))
+⟨(λ ⟨t, _, hp⟩, ⟨t, _, hp⟩), (λ ⟨t, hs, hp⟩, ⟨t, mem_powerset.2 hs, hp⟩)⟩
+
+/-- For predicate `p` decidable on subsets, it is decidable whether `p` holds for every subset. -/
+def decidable_forall_of_decidable_subsets {s : finset α} {p : ∀ t ⊆ s, Prop}
+  (hu : ∀ t (h : t ⊆ s), decidable (p t h)) : decidable (∀ t (h : t ⊆ s), p t h) :=
+decidable_of_iff (∀ t (h : t ∈ s.powerset), p t (mem_powerset.1 h))
+⟨(λ h t hs, h t (mem_powerset.2 hs)), (λ h _ _, h _ _)⟩
+
+variables (p : finset α → Prop) [decidable_pred p]
+
+instance decidable_exists_of_subsets (s : finset α) : decidable (∃ t ⊆ s, p t) :=
+decidable_exists_of_decidable_subsets infer_instance
+
+instance decidable_forall_of_subsets (s : finset α) : decidable (∀ t ⊆ s, p t) :=
+decidable_forall_of_decidable_subsets infer_instance
```

Comment:
I'm surprised you need to provide lean with these instances - doesn't it already know how to solve `∀ t ⊆ s, p t` (aka `∀ t (H : t ⊆ s), p t`based on the fact that `t` is a `fintype` and `t ⊆ s` should be decidable?

Reply:
`t` here would be a `finset α` which might not be a fintype in general, it's essential that it's a subset of `s` (and so there are finitely many choices, which is exactly what this proof does via `decidable_forall_of_decidable_subsets`

Reply:
Well `fintype` is new to me but I've had a quick look at it just now. Even if I include `decidable_eq α` to make `t ⊆ s` decidable, it doesn't seem to solve this and I don't (yet) understand it well enough to guess by what mechanism it might...

Reply:
Thanks @b-mehta, I thought I was probably missing something.